### PR TITLE
Feat/chat

### PR DESCRIPTION
**Pull request by MarcusSchweizer91**
chat frontend und backend

### CODE REVIEW

File: frontend/src/CardGallery/components/Chat/UserChat.tsx [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    const [message, setMessage] = useState('');
+    const [ws, setWs] = useState<WebSocket | null>(null);
+    const { senderUsername, receiverUsername } = useParams<{ senderUsername: string, receiverUsername: string }>();
+
+    console.log(messages);
+    useEffect(() => {
+        const ws = new WebSocket(`ws://localhost:8080/api/ws/chat`);
+        setWs(ws);
+
+        ws.onmessage = (event) => {
```
Comment (MoinMarcell) [inline frontend/src/CardGallery/components/Chat/UserChat.tsx:17]: What type of Event? It can be any type I guess?

**Review by MoinMarcell — APPROVED** [review]
lgtm, one question below :)

File: frontend/src/CardGallery/components/Chat/UserChat.tsx (hunk 1 above)
Comment (MarcusSchweizer91) [reply frontend/src/CardGallery/components/Chat/UserChat.tsx:17]: ja ich glaube schon

File: backend/src/main/java/com/example/backend/controller/ExchangeController.java [hunk 2]
```diff
@@ -36,8 +37,8 @@ public ExchangeCard getByID(@PathVariable String id) {
     }
 
     @PutMapping(path = "/{id}")
-    public ExchangeCard updateEntry(@PathVariable String id, @RequestPart(value = "entry") String entryToUpdate, @RequestPart(value = "file") MultipartFile cardImage) throws IOException {
-        return exchangeService.updateEntry(id, entryToUpdate, cardImage);
+    public ExchangeCard updateEntry(@PathVariable String id, @RequestPart(value = "entry") String entryToUpdate, @RequestPart(value = "file") MultipartFile cardImage, String author) throws IOException {
```
Comment (Flooooooooooorian) [inline backend/src/main/java/com/example/backend/controller/ExchangeController.java:40]: Where does the author Parameter come from? There ist no @RequestBody or any other Annotation?

File: backend/src/main/java/com/example/backend/service/ChatService.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    private final Set<WebSocketSession> session = new HashSet<>();
+
+    public ChatService(ChatRepo chatRepo, ObjectMapper objectMapper) {
+        this.chatRepo = chatRepo;
+        this.objectMapper = objectMapper;
+    }
+
+    public Set<WebSocketSession> getSession(){
+        return session;
+    }
```
Comment (Flooooooooooorian) [inline backend/src/main/java/com/example/backend/service/ChatService.java:30]: Why do you need this Method?

File: backend/src/test/java/com/example/backend/service/ChatServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+        ChatService spyChatService = spy(new ChatService(chatRepo, objectMapper));
+        Set<WebSocketSession> expectedSessions = new HashSet<>();
+        doReturn(expectedSessions).when(spyChatService).getSession();
+
+
+        Set<WebSocketSession> actualSessions = spyChatService.getSession();
+
+
+        assertEquals(expectedSessions, actualSessions);
+    }
```
Comment (Flooooooooooorian) [inline backend/src/test/java/com/example/backend/service/ChatServiceTest.java:57]: What exactly do you test here? You mock the getSession Method... So you are testing you mock?

File: backend/src/test/java/com/example/backend/service/ChatServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+                        .timestamp(LocalDateTime.now())
+                        .build(),
+                ChatMessage.builder()
+                        .senderUsername("receiver")
+                        .receiverUsername("sender")
+                        .message("Hi")
+                        .timestamp(LocalDateTime.now())
+                        .build()
+        );
+        lenient().when(session.getPrincipal()).thenReturn(() -> "sender");
```
Comment (Flooooooooooorian) [inline backend/src/test/java/com/example/backend/service/ChatServiceTest.java:89]: Why do you use `lenient()` here?

File: backend/src/test/java/com/example/backend/service/ChatServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+        assertEquals("{\"sender\":\"receiver\",\"receiver\":\"sender\",\"message\":\"Hi\",\"timestamp\":\"" + messages.get(1).getTimestamp() + "\"}", capturedTextMessages.get(1).getPayload());
+        verify(chatRepo).findAllBySenderUsernameAndReceiverUsername("sender", "receiver");
+        verify(chatRepo).findAllBySenderUsernameAndReceiverUsername("receiver", "sender");
+        verifyNoMoreInteractions(session, chatRepo, objectMapper);
+    }
+
+    @Test
+    void afterConnectionEstablished_shouldSendPreviousMessages() throws Exception {
+
+        ChatService chatServiceSpy = spy(chatService);
```
Comment (Flooooooooooorian) [inline backend/src/test/java/com/example/backend/service/ChatServiceTest.java:109]: Why are you using a spy here?

File: frontend/src/CardGallery/components/Chat/UserChat.tsx [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+import {ChatMessage, NewChatMessage} from "../../models/ChatMessage";
+
+export default function UserChat(){
+
+    const [messages, setMessages] = useState<ChatMessage[]>([]);
+    const [message, setMessage] = useState('');
+    const [ws, setWs] = useState<WebSocket | null>(null);
+    const { senderUsername, receiverUsername } = useParams<{ senderUsername: string, receiverUsername: string }>();
+
+    console.log(messages);
```
Comment (Flooooooooooorian) [inline frontend/src/CardGallery/components/Chat/UserChat.tsx:12]: Remove console.log

File: frontend/src/CardGallery/components/Chat/UserChat.tsx [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    useEffect(() => {
+        const ws = new WebSocket(`ws://localhost:8080/api/ws/chat`);
+        setWs(ws);
+
+        ws.onmessage = (event) => {
+            const data = JSON.parse(event.data);
+            console.log(event);
```
Comment (Flooooooooooorian) [inline frontend/src/CardGallery/components/Chat/UserChat.tsx:19]: Remove console.log

**Review by Flooooooooooorian — COMMENTED** [review]
Habe jetzt schnell drüber geguckt und dir ein paar Kommentare da gelassen :)

**Pull request merged**